File: Sources/CompanionSessionCore/WAVEncoder.swift
import Foundation

/// Wraps raw little-endian PCM in a canonical 44-byte RIFF/WAVE header so the
/// uploaded audio is self-describing.
///
/// The companion captures headerless PCM sidecars (mic = mono, system = stereo).
/// Uploading them raw forces the backend to *guess* the format — and its guess
/// (always stereo) silently corrupts the mono mic channel. Prepending an
/// accurate header removes the guess entirely: the backend passes any `RIFF`
/// payload straight through to transcription.
public enum WAVEncoder {
    /// Prepends a WAV header describing `pcm` as `channels`-channel,
    /// `bitsPerSample`-bit, `sampleRate`-Hz little-endian PCM.
    public static func wrap(
        pcm: Data,
        sampleRate: Int,
        channels: Int,
        bitsPerSample: Int = 16
    ) -> Data {
        var out = header(
            dataByteCount: pcm.count,
            sampleRate: sampleRate,
            channels: channels,
            bitsPerSample: bitsPerSample
        )
        out.append(pcm)
        return out
    }

    /// The 44-byte canonical RIFF/WAVE header for a `data` chunk of exactly
    /// `dataByteCount` bytes.
    ///
    /// Split out from ``wrap(pcm:sampleRate:channels:bitsPerSample:)`` so a
    /// large PCM sidecar can be turned into a WAV on disk without ever holding
    /// the samples in memory: write this header, then stream the PCM bytes in
    /// after it (see the signed-URL upload path, which must not buffer a
    /// multi-hundred-MB session).
    public static func header(
        dataByteCount: Int,
        sampleRate: Int,
        channels: Int,
        bitsPerSample: Int = 16
    ) -> Data {
        let byteRate = sampleRate * channels * bitsPerSample / 8
        let blockAlign = channels * bitsPerSample / 8
        let dataSize = UInt32(dataByteCount)
        let chunkSize = 36 + dataSize

        var header = Data(capacity: 44)
        header.append(contentsOf: Array("RIFF".utf8))
        appendUInt32LE(chunkSize, to: &header)
        header.append(contentsOf: Array("WAVE".utf8))
        header.append(contentsOf: Array("fmt ".utf8))
        appendUInt32LE(16, to: &header) // PCM fmt chunk size
        appendUInt16LE(1, to: &header) // audio format = PCM
        appendUInt16LE(UInt16(channels), to: &header)
        appendUInt32LE(UInt32(sampleRate), to: &header)
        appendUInt32LE(UInt32(byteRate), to: &header)
        appendUInt16LE(UInt16(blockAlign), to: &header)
        appendUInt16LE(UInt16(bitsPerSample), to: &header)
        header.append(contentsOf: Array("data".utf8))
        appendUInt32LE(dataSize, to: &header)
        return header
    }

    private static func appendUInt32LE(_ value: UInt32, to data: inout Data) {
        var le = value.littleEndian
        Swift.withUnsafeBytes(of: &le) { data.append(contentsOf: $0) }
    }

    private static func appendUInt16LE(_ value: UInt16, to data: inout Data) {
        var le = value.littleEndian
        Swift.withUnsafeBytes(of: &le) { data.append(contentsOf: $0) }
    }
}

File: windows/PabloCompanion.Core/WAVEncoder.cs
using System.Buffers.Binary;
using System.Text;

namespace PabloCompanion.Core;

/// <summary>
/// Builds the canonical 44-byte RIFF/WAVE header that makes uploaded PCM
/// self-describing.
///
/// The companion captures headerless PCM sidecars (mic = mono, system = stereo).
/// Uploading them raw under an <c>audio/wav</c> mime forces the backend to
/// <i>guess</i> the format — and its guess (always stereo) silently corrupts the
/// mono mic channel: frames halve and the waveform mangles, so the therapist
/// audio transcribes to nothing. Prepending an accurate header removes the guess
/// entirely; the backend passes any <c>RIFF</c> payload straight through.
///
/// The C# mirror of the macOS <c>WAVEncoder</c> in <c>CompanionSessionCore</c> —
/// both platforms must produce byte-identical headers, so change them together.
/// </summary>
public static class WAVEncoder
{
    /// <summary>Size of the canonical PCM header this encoder emits.</summary>
    public const int HeaderSize = 44;

    /// <summary>
    /// Builds a 44-byte header describing <paramref name="dataByteCount"/> bytes of
    /// little-endian PCM. Split from <see cref="Wrap"/> so a caller can stream the
    /// payload: an hour of 48 kHz audio is hundreds of megabytes, which shouldn't
    /// have to sit in memory just to be prefixed.
    /// </summary>
    /// <param name="dataByteCount">Byte length of the PCM that will follow the header.</param>
    /// <param name="sampleRate">Frames per second, e.g. 48000.</param>
    /// <param name="channels">1 for the mono mic, 2 for stereo system audio.</param>
    /// <param name="bitsPerSample">Bits per sample; the capture path emits 16.</param>
    public static byte[] BuildHeader(long dataByteCount, int sampleRate, int channels, int bitsPerSample = 16)
    {
        ArgumentOutOfRangeException.ThrowIfNegative(dataByteCount);
        ArgumentOutOfRangeException.ThrowIfLessThan(sampleRate, 1);
        ArgumentOutOfRangeException.ThrowIfLessThan(channels, 1);
        ArgumentOutOfRangeException.ThrowIfLessThan(bitsPerSample, 1);

        // RIFF sizes are unsigned 32-bit; past 4 GiB the format simply can't
        // describe the payload, and silently truncating would corrupt it.
        var chunkSize = 36 + dataByteCount;
        ArgumentOutOfRangeException.ThrowIfGreaterThan(chunkSize, uint.MaxValue, nameof(dataByteCount));

        var byteRate = sampleRate * channels * bitsPerSample / 8;
        var blockAlign = channels * bitsPerSample / 8;

        var header = new byte[HeaderSize];
        var span = header.AsSpan();

        WriteAscii(span[..4], "RIFF");
        BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(4, 4), (uint)chunkSize);
        WriteAscii(span.Slice(8, 4), "WAVE");
        WriteAscii(span.Slice(12, 4), "fmt ");
        BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(16, 4), 16); // PCM fmt chunk size
        BinaryPrimitives.WriteUInt16LittleEndian(span.Slice(20, 2), 1);  // audio format = PCM
        BinaryPrimitives.WriteUInt16LittleEndian(span.Slice(22, 2), (ushort)channels);
        BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(24, 4), (uint)sampleRate);
        BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(28, 4), (uint)byteRate);
        BinaryPrimitives.WriteUInt16LittleEndian(span.Slice(32, 2), (ushort)blockAlign);
        BinaryPrimitives.WriteUInt16LittleEndian(span.Slice(34, 2), (ushort)bitsPerSample);
        WriteAscii(span.Slice(36, 4), "data");
        BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(40, 4), (uint)dataByteCount);

        return header;
    }

    /// <summary>
    /// Prepends a WAV header to <paramref name="pcm"/>, returning the complete file
    /// bytes. Convenient for small payloads and tests; prefer
    /// <see cref="BuildHeader"/> plus a stream copy for real recordings.
    /// </summary>
    public static byte[] Wrap(byte[] pcm, int sampleRate, int channels, int bitsPerSample = 16)
    {
        ArgumentNullException.ThrowIfNull(pcm);

        var header = BuildHeader(pcm.Length, sampleRate, channels, bitsPerSample);
        var output = new byte[header.Length + pcm.Length];
        header.CopyTo(output, 0);
        pcm.CopyTo(output, header.Length);
        return output;
    }

    /// <summary>
    /// Whether these bytes already begin a RIFF container, in which case they are
    /// self-describing and must be uploaded untouched rather than double-wrapped.
    /// </summary>
    public static bool IsRiff(ReadOnlySpan<byte> data) =>
        data.Length >= 4 && data[0] == (byte)'R' && data[1] == (byte)'I'
                         && data[2] == (byte)'F' && data[3] == (byte)'F';

    private static void WriteAscii(Span<byte> destination, string value) =>
        Encoding.ASCII.GetBytes(value, destination);
}
